### PR TITLE
Feature TR-3070 delivery property patch api

### PR DESCRIPTION
**Pull request by wazelin**
# [TR-3070](https://oat-sa.atlassian.net/browse/TR-3070)

## Dependencies
- https://github.com/oat-sa/tao-core/pull/3322

## How to test
Import [CG Delivery.postman_collection.json.zip](https://github.com/oat-sa/extension-tao-delivery-rdf/files/7871100/CG.Delivery.postman_collection.json.zip) into Postman.
You may run it against a test environment as mention [here](https://oat-sa.atlassian.net/browse/TR-3070?focusedCommentId=168634).

https://user-images.githubusercontent.com/2943256/149480109-d07487bb-1d04-442e-a1f0-6d8ce4b79b3d.mov

## Changelog
- feat: introduce `DeliveryServiceSearchRequest` and its validator and handler
- feat: implement delivery property PATCH request handlers
- chore: transform `rest` open API doc from `2.0` to `3.0` version
- feat: implement delivery property PATCH request handlers
- chore: bump `tao-core` dependency to the one implementing extended HTTP status code handling
- fix: simplify excludede user ID comparison in `DeliveryMgmt::excludeTesttaker()`
- feat: introduce `DeliveryFormFactory` to facilitate `DeliveryForm` instantiation
- feat: implement `taoDeliveryRdf/RestDelivery/updateProperties` partial Delivery property PATCH request handler

### CODE REVIEW

File: model/Delivery/Presentation/Web/RequestHandler/DeliveryPatchRequestHandler.php [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        foreach ($this->patchRequestHandlers as $patchRequestHandler) {
+            if ($patchRequestHandler->isApplicable($request)) {
+                return new DeliveryUpdateInput(
+                    $searchRequest,
+                    $patchRequestHandler->handle($request)
+                );
+            }
+        }
+
+        throw new BadRequestException('Unsupported Media Type', 415);
```
Comment (yohanlaborda) [inline model/Delivery/Presentation/Web/RequestHandler/DeliveryPatchRequestHandler.php:63]: `415` I don't know what the number means, it is better to create a constant

File: model/Delivery/Presentation/Web/RequestHandler/DeliverySearchRequestHandler.php [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+     * @throws BadRequestException
+     */
+    public function handle(ServerRequestInterface $request): DeliverySearchRequest
+    {
+        $this->validator->validate($request);
+
+        $queryParameters = $request->getQueryParams();
+
+        return new DeliverySearchRequest(
+            $queryParameters[self::ID]
```
Comment (yohanlaborda) [inline model/Delivery/Presentation/Web/RequestHandler/DeliverySearchRequestHandler.php:56]: What if the identification does not exist?

File: model/Delivery/ServiceProvider/DeliveryServiceProvider.php [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        $services
+            ->set(DeliveryService::class, DeliveryService::class)
+            ->public()
+            ->args([
+                service(DeliveryRepository::class),
+                service(DeliveryFormFactory::class),
+                service(EventManager::class),
+            ]);
+
+        $services
```
Comment (yohanlaborda) [inline model/Delivery/ServiceProvider/DeliveryServiceProvider.php:58]: In order not to create a large method, you can create a short method with the definition of each service

File: test/unit/model/Delivery/DataAccess/DeliveryRepositoryTest.php [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    private $deliveryClassMock;
+
+    /**
+     * @before
+     */
+    public function init(): void
+    {
+        $this->deliveryClassMock = $this->createMock(KernelClass::class);
+
+        $this->sut = $this->createPartialMock(DeliveryRepository::class, ['getClass']);
```
Comment (yohanlaborda) [inline test/unit/model/Delivery/DataAccess/DeliveryRepositoryTest.php:51]: What is `sut`?

**Review by yohanlaborda — CHANGES_REQUESTED** [review]

File: controller/RestDelivery.php [hunk 5]
```diff
@@ -187,13 +196,38 @@ public function update()
...
+            $delivery = $this->getDeliveryService()->update(
+                $this->getDeliveryPatchRequestHandler()->handle(
+                    $this->getPsrRequest()
+                )
+            );
+        } catch (RuntimeException $exception) {
+            throw new BadRequestException($exception->getMessage(), $exception->getCode());
+        }
+
+        /** @noinspection PhpUnhandledExceptionInspection */
```
[the remark this answers is not in the document]
Comment (wazelin) [reply controller/RestDelivery.php:220]: That method will never thought the exception it specifies in this context.

File: model/Delivery/Business/Service/DeliveryService.php [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+     */
+    private function setProperties(DeliveryUpdateInput $input, Form $form): void
+    {
+        $errors = [];
+
+        foreach ($input->getProperties() as $property => $value) {
+            $formElement = $form->getElement(UriHelper::encode($property));
+
+            if (null === $formElement) {
+                continue;
```
[the remark this answers is not in the document]
Comment (wazelin) [reply model/Delivery/Business/Service/DeliveryService.php:96]: Why? Early return makes the less nested code more readable.

File: model/Delivery/Presentation/Web/RequestHandler/DeliverySearchRequestHandler.php (hunk 2 above)
Comment (wazelin) [reply model/Delivery/Presentation/Web/RequestHandler/DeliverySearchRequestHandler.php:56]: Then we'll have a runtime error here.
Such a condition is prevented by the request validator layer, though.

File: model/Delivery/Presentation/Web/RequestHandler/DeliveryPatchRequestHandler.php (hunk 1 above)
Comment (wazelin) [reply model/Delivery/Presentation/Web/RequestHandler/DeliveryPatchRequestHandler.php:63]: I don't like neither of those approaches to be fair.

Ideally, every single HTTP-related exception should have its own implementation.
Until we have them, I'd prefer not to create additional http status code storing classes.

Comment (yohanlaborda) [reply model/Delivery/Presentation/Web/RequestHandler/DeliveryPatchRequestHandler.php:63]: We can add the constant in the class itself to make it more explicit

**Review by yohanlaborda — APPROVED** [review]
- [ ] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [X] New code is respecting code style rules
- [X] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [X] Acceptance criteria are respected
- [X] Pull request title and description are meaningful
- [X] Pull request's target is not master
- [X] Commits are following conventional commits
- [ ] Changelog is updated according to changes (if applicable)
- [ ] Documentation is updated according to changes (if applicable)

**Review by poyuki — APPROVED** [review]
- [x] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x] New code is respecting code style rules
- [x] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [x] Pull request title and description are meaningful
- [x] Pull request's target is not `master`
- [x] Commits are following [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/)
- [ ] Changelog is updated according to changes (if applicable)
- [ ] Documentation is updated according to changes (if applicable)

File: composer.json [hunk 7]
```diff
@@ -22,7 +22,7 @@
     "require": {
         "oat-sa/oatbox-extension-installer": "~1.1||dev-master",
         "oat-sa/generis": ">=14.0.0",
-        "oat-sa/tao-core": ">=48.58.0",
+        "oat-sa/tao-core": ">=48.80.0",
```
Comment (wazelin) [inline composer.json:25]:
```suggestion
        "oat-sa/tao-core": ">=48.81.0",
```

**Pull request merged**